Python Wheel installation guide for AC (#280)
astronomer/docs: ac/next/install-vm.md

---
title: "Install Apache Airflow on a Virtual Machine"
navTitle: "Install on a Virtual Machine"
description: "Install Apache Airflow on one or more virtual machines with the Astronomer Core Python wheel."
---

## Overview

This guide walks through the necessary steps to install Apache Airflow via the Astronomer Core Python wheel on one or more Virtual Machines (VMs). By the end of the setup, you'll be able to deploy and run Airflow across multiple machines.

> 

Author: If you're running Ubuntu, you don't need to run a Virtual Machine software to emulate another OS. The virtual machine here is what you create in `C. Create a virtual environment`. While people might run a VM to access an Ubuntu environment, it's unlikely they'd be doing that for a production-scale deployment. In reality, they'd probably already have Linux-based machines or servers upon which they mount the `venv` we create here.

> 

Reviewer: Hm, ok - might revisit this but sounds good for now.

If you haven't tested Airflow locally and would like to do so, refer to [Astronomer Core Quickstart](/docs/ac/next/01_quickstart).

## Prerequisites

First, ensure the OS-level packages listed below are installed on your machines. If you're Debian-based, run `$ apt get <package>` to do so. If you're running RedHat Linux, run `$ yum install <package>`.

- sudo
- python3
- python3-dev (python3-devel for RHEL/CentOS)
- gcc

> You also need a database that is reachable by all the machines that will be running this Airflow cluster. While this guide walks through the process for configuring a PostgreSQL database, Airflow is compatible with all of the following databases:

Author: Given that we create a virtual environment later, and people could mistake that for what we're referencing here (Airflow as a whole, not the actual VM running it), I like instance.

> If your database server is running on the same machine as your other Airflow components, you can change your authentication method from `peer` to `md5` in the same `pg_hba.conf` file to allow connections with a username/password from the same machine.

### Alternative setup: Use an existing database

Instead of creating a new PostgreSQL database, you can use an existing database as long as both of the following are true:

- The database is compatible with Airflow as described in Prerequisites.
- A user named `airflow` has ownership access to the database.

When you specify the `AIRFLOW__CORE__SQL_ALCHEMY_CONN` environment variable in step 2F, replace the connection URL with the appropriate URL for your database.

## Step 2: Configure Each Machine in Your System

Complete the steps below for each machine that will be running a core Apache Airflow component (Scheduler, Webserver, or Worker).

### A. Create a system user to run Airflow

Airflow can run as any user, but for this setup we assume the user name `astro`. Run the following command to add this user to your machine:

```sh
sudo useradd --shell=/bin/false --create-home astro
```

### B. Create an Airflow project directory

> You also need to configure an AIRFLOW_HOME directory (not to be confused with the user's home directory) where you'll store your DAGs. We recommend using the path `/usr/local/airflow` as your project directory and `/usr/local/airflow/dags` as your DAG directory, but any path can be chosen as long as the `astro` user has write access to it. To do this, run the following commands:

Reviewer: ```suggestion
You also need to configure an `AIRFLOW_HOME` directory (not to be confused with the user's home directory) where you'll store your DAGs. We recommend using the path `/usr/local/airflow` as your project directory and `/usr/local/airflow/dags` as your DAG directory, but any path can be chosen as long as the `astro` user has write access to it. To do this, run the following commands:
```

We have `AIRFLOW_HOME` ticked before, so we should just make sure to stay consistent

```sh
sudo install --owner=astro --group=astro -d /usr/local/airflow
echo 'export AIRFLOW_HOME=/usr/local/airflow' | sudo tee --append ~astro/.bashrc
cd /usr/local/airflow && mkdir dags
```

> **Note:** If you're running Airflow on multiple machines, each machine needs access to the same DAGs in order to successfully execute them. We recommend setting up automation pipelines for updating all of your DAG folders whenever a local folder is updated. For more information, read [Deploying DAGs].

### C. Create a virtual environment

> To isolate the Astronomer Core Python modules from changes to the system, create a virtual environment using the following command:

Author: Where else do we refer to VMs/VEs in this way? I think people will understand what we mean without the parenthetical phrase here.

```sh
sudo -u astro python3 -m venv ~astro/airflow-venv
```

> 

Reviewer: `venv` is a tool to create lightweight, isolated Python environments that allow you to install modules and packages without affecting systemwide configuration. This command will create a virtual Python environment in a directory named `astro/airflow-venv`. For more information, read [Python's`venv` documentation](https://docs.python.org/3/library/venv.html).

### D. Install Astronomer Core

Install the AC Python wheel onto your machine by running:

```sh
sudo -u astro ~astro/airflow-venv/bin/pip install --extra-index-url=https://pip.astronomer.io/simple/ 'astronomer-core[postgres]==<airflow-version>'
```

To install the latest patch version of Apache Airflow 2.0.0, for example, this command would be:

```sh
sudo -u astro ~astro/airflow-venv/bin/pip install --extra-index-url=https://pip.astronomer.io/simple/ 'astronomer-core[postgres]==2.0.0.x'
```

> This command includes the optional `[postgres]` dependency so that all libraries needed to use Postgres are also installed. If you are using a different database or require additional dependencies, specify those dependencies in a comma-delimited list (for example: `[mysql, redis, crypto, aws, celery]`). For a list of all optional dependencies, refer to the [AC pip index](https://pip.astronomer.io/simple/index.html).

Reviewer: ```suggestion
This command includes the optional `[postgres]` dependency so that all libraries needed to use Postgres are also installed. If you are using a different database or require additional dependencies, specify those dependencies in a comma-delimited list. For example, you might add: `[mysql, redis, crypto, aws, celery]`. For a list of all optional dependencies, refer to the [AC pip index](https://pip.astronomer.io/simple/index.html).
```

Stupid q - Is this the postgres airflow provider, or just the equivalent of running `pip install postgres`?

Author: I think both? You can either install the dependencies later, per our guide, or install them now using this shorthand syntax. It's the same software either way.

Author: Also moved some of this pgraph around to make the example more clear.

### E. Configure a process supervisor

To ensure that Airflow is always running when your machine is on, we recommend implementing a process supervisor. [Systemd](https://systemd.io/) is used in this example, though any process supervisor will work here.

To use systemd as a process supervisor:

1. Create a systemd unit file using the following command:

    ```sh
    sudo -e /etc/systemd/system/astronomer-core@.service
    ```

> 2. Using a text editor, create a file called `astronomer-core` and edit it to contain these environment variables and values:

Reviewer: Should this file be within the `/usr/local/airflow` directory, or does it not really matter?

Author: I don't think it matters, since the path to the virtual environment is defined in the file.

    ```sh
    AIRFLOW_HOME=/usr/local/airflow/
    AIRFLOW__CORE__LOAD_EXAMPLES=False
    PATH=$PATH:/home/astro/airflow-venv/bin
    ```

    When you run Airflow for the first time, a file called `airflow.cfg` will be generated in your `AIRFLOW_HOME` directory. If you want to configure environment variables that apply to all of your machines, we recommend specifying them in that `airflow.cfg` file. For more information, read the Apache Airflow documentation on [Setting Configuration Options](https://airflow.apache.org/docs/apache-airflow/stable/howto/set-config.html).

> 3. Using the text editor of your choice, add the following to your systemd unit file:

Reviewer: ```suggestion
3. Now, add the following to the systemd unit file you created in step 1:
```

("Using the text editor of your choice" felt redundant given the start of step 2)

Reviewer: Separate comment/question - I can see users adding this stuff in step 3 to the `astronomer-core` file since they just created it. If it doesn't matter, should we swap steps 2 and 3? And then in step 3 they create `astronomer-core`, and it may flow more easily to step 2F, where they need to add env vars to that file

Author: Would that flow better? In step 3, they have to specify the file they create in step 2, such that they wouldn't know the file path at the time of configuring it. I'm also hesitant to mention "Step 1" here lest people think it's the header actually called "Step 1".

Reviewer: Think we decided the order here was the best we could do

    ```
    [Unit]
    Description=Airflow %I daemon
    After=network-online.target cloud-config.service
    Requires=network-online.target

    [Service]
    EnvironmentFile=/etc/default/astronomer-core
    User=astro
    Group=astro
    Type=simple
    WorkingDirectory=/usr/local/airflow/
    ExecStart=/home/astro/airflow-venv/bin/airflow %i
    Restart=always
    RestartSec=5s

    [Install]
    WantedBy=multi-user.target

>     ```

Reviewer: Would be good to have an explanation of what this all does 🙃

Author: I think combined with the description at the start of these steps, these variables and their functions are self-evident enough. If users want to know more, they can always go to the Systemd website linked above and read about configurations there.

### F. Configure Airflow for Database Access

To connect your Airflow environment to the metadata DB you created in Step 1, add the following environment variables to your `astronomer-core` file:

- For Local Executor:

    ```
    AIRFLOW__CORE__SQL_ALCHEMY_CONN=postgresql://airflow:<your-user-password>@localhost/airflow
    AIRFLOW__WEBSERVER__BASE_URL=http://host:port
    AIRFLOW__CORE__EXECUTOR=LocalExecutor
    ```

- For Celery Executor:

    ```
    AIRFLOW__CORE__SQL_ALCHEMY_CONN=postgresql://airflow:<your-user-password>@localhost/airflow
    AIRFLOW__WEBSERVER__BASE_URL=http://host:port
    AIRFLOW__CELERY__BROKER_URL=sqla+postgresql://airflow:<your-user-password>@localhost/airflow
    AIRFLOW__CELERY__RESULT_BACKEND=db+postgresql://airflow:<your-user-password>@localhost/airflow
    AIRFLOW__CORE__EXECUTOR=CeleryExecutor
    ```

The password you specify here should be the same one you specified when prompted by the `createuser` command in Step 1. If your password contains `%`, `/`, or `@` then you will need to url-escape; replace `%` with `%25`, `/` with `%2F`, and `@` with `%40`.

> 

Reviewer: Maybe add 1 sentence describing these executors, and link out to https://www.astronomer.io/guides/airflow-executors-explained? Could steal from https://www.astronomer.io/docs/cloud/stable/deploy/configure-deployment#select-an-executor

Author: Should we link out to these resources if we don't have anything to say about KubeExecutor yet 🤒

* In this example setup, the `AIRFLOW__CELERY__BROKER_URL` environment variable uses the main database for communication. To take some load off of your main database, we recommend using a dedicated message broker such as [Redis](https://redis.io/) and specifying that here instead.

## Step 3: Set Up the Scheduler

In Airflow, [the Scheduler](https://airflow.apache.org/docs/apache-airflow/stable/scheduler.html) is responsible for reading from the metadata database to check on the status of each task and decides the order in which tasks should be completed. To get your Scheduler running:  

1. Enable the Scheduler by running the following command:

    ```sh
    sudo systemctl enable astronomer-core@scheduler.service
    ```

2. Edit the override file for the machine by running the following command:

    ```sh
    sudo systemctl edit astronomer-core@scheduler.service
    ```

3. In the empty document that appears, add the following lines:

    ```
    [Service]
    ExecStartPre=/home/astro/airflow-venv/bin/airflow upgradedb
    ```

4. Start the service by running:   

    ```sh
    sudo systemctl start astronomer-core@scheduler.service
    ```

## Step 4: Set Up the Webserver

> The Webserver is an Airflow core component that is responsible for rendering the Airflow UI. To configure it on its own machine, follow the steps below.

Reviewer: ```suggestion
[The Webserver](https://airflow.apache.org/docs/apache-airflow/stable/security/webserver.html) is a core Airflow component that is responsible for rendering the Airflow UI. To configure it on its own machine, follow the steps below.
```

1. Enable the Webserver by running the following:

    ```sh
    sudo systemctl enable astronomer-core@webserver.service
    ```

2. Start the Webserver by running the following:

    ```sh
    sudo systemctl start astronomer-core@webserver.service
    ```

> You can now access the Airflow UI in a web browser via `http://host:port`.

Reviewer: Is there a default `host:port`? Without any customization - did they set this prior?

Author: The default is `localhost:8080` as elsewhere, but we tell them to reconfigure it as an environment variable above. This is what we say to do in our existing guides, but I also don't see the harm in leaving it as the default value. What do you think?

> **Note:** For added security and stability, we recommend running the Webserver behind a reverse proxy and load balancer such as [nginx](https://www.nginx.com/). For more information on this feature, read the [Apache Airflow documentation](https://airflow.apache.org/docs/stable/howto/run-behind-proxy.html).

## Step 5: Set Up the Worker Machines

For each machine on which you want to host a Worker:

1. Enable the Worker service by running the following command:

   ```sh
   sudo systemctl enable astronomer-core@worker.service
   ```

2. Start the service by running the following command:

    ```sh
    sudo systemctl start astronomer-core@worker.service
    ```

You now have the ability to run Airflow tasks within DAGs.

> ## Step 6: Confirm the Installation

Reviewer: ```suggestion
## Step 6: Confirm your Airflow Installation
```

Author: Not sure what additional context this adds that's not present in the original. In this case, we're not referring to Airflow as "installations" like some Enterprise folks refer to it as, so not sure why we'd use a possessive here.

> To confirm that the installation was successful, open `http://host:port` in your web browser. You should see the login screen for the Airflow UI. Log in with `admin` as both your user name and password. From there, you should see Airflow's primary 'DAGs' view:

Reviewer: ```suggestion
To confirm that you successfully installed Apache Airflow, open `http://host:port` in your web browser. You should see the login screen for the Airflow UI. Log in with `admin` as both your user name and password. From there, you should see Airflow's primary 'DAGs' view:
```

![Empty Airflow UI](https://assets2.astronomer.io/main/docs/airflow-ui/installation-home.png)

If you want to further confirm that everything's working as intended, add this [example DAG] to the DAG folder of every machine running Airflow, then restart your Airflow services using the following commands:

```sh
sudo systemctl stop astronomer-core@webserver.service && sudo systemctl start astronomer-core@webserver.service
sudo systemctl stop astronomer-core@scheduler.service && sudo systemctl start astronomer-core@scheduler.service
sudo systemctl stop astronomer-core@worker.service && sudo systemctl start astronomer-core@worker.service
```

> When you reopen the Airflow UI, you should see the example DAG ready to run.

Reviewer: ```suggestion
When you refresh the Airflow UI, you should see the example DAG ready to run.
```

Assume more accurate?

* In this example setup, the `AIRFLOW__CELERY__BROKER_URL` environment variable uses the main database for communication. To take some load off of your main database, we recommend using a dedicated message broker such as [Redis](https://redis.io/) and specifying that here instead.

## Step 3: Set Up the Scheduler

In Airflow, [the Scheduler](https://airflow.apache.org/docs/apache-airflow/stable/scheduler.html) is responsible for reading from the metadata database to check on the status of each task and decides the order in which tasks should be completed. To get your Scheduler running:  

1. Enable the Scheduler by running the following command:

    ```sh
    sudo systemctl enable astronomer-core@scheduler.service
    ```

2. Edit the override file for the machine by running the following command:

    ```sh
    sudo systemctl edit astronomer-core@scheduler.service
    ```

3. In the empty document that appears, add the following lines:

    ```
    [Service]
    ExecStartPre=/home/astro/airflow-venv/bin/airflow upgradedb
    ```

4. Start the service by running:   

    ```sh

>     sudo systemctl start astronomer-core@scheduler.service

Author: Got the following error when doing this:
```
Process: 36399 ExecStartPre=/home/astro/airflow-venv/bin/airflow db upgrade (code=exited, status=1/FAILURE)

d[1]: astronomer-core@scheduler.service: Failed with result 'exit-code'.
d[1]: Failed to start Airflow scheduler daemon.
```

Reviewer: ^Curious why you got this error 👀  If it could be a common user mistake, do you think something worth including in a ` > **Note**` or nah?

Author: Oh this is super old and definitely was just because I had my configurations all messed up.

>
> If your database server is running on the same machine as your other Airflow components, you can change your authentication method from `peer` to `md5` in the same `pg_hba.conf` file to allow connections with a username/password from the same machine.

### Alternative setup: Use an existing database

Instead of creating a new PostgreSQL database, you can use an existing database as long as both of the following are true:

- The database is compatible with Airflow as described in Prerequisites.
- A user named `airflow` has ownership access to the database.

When you specify the `AIRFLOW__CORE__SQL_ALCHEMY_CONN` environment variable in step 2F, replace the connection URL with the appropriate URL for your database.

## Step 2: Configure Each Machine in Your System

Complete the steps below for each machine that will be running a core Apache Airflow component (Scheduler, Webserver, or Worker).

### A. Create a system user to run Airflow

Airflow can run as any user, but for this setup we assume the user name `astro`. Run the following command to add this user to your machine:

```sh

> sudo useradd --shell=/bin/false --create-home astro

Author: Remove --shell=/bin/false

Author: Astro user did not have permission on the environment file.

---
title: "Install Apache Airflow on a Virtual Machine"
navTitle: "Install on a Virtual Machine"
description: "Install Apache Airflow on one or more virtual machines with the Astronomer Core Python wheel."
---

## Overview

This guide contains the complete setup steps for installing Apache Airflow via the Astronomer Core Python wheel on one or more Virtual Machines (VMs). By the end of the setup, you'll be able to deploy and run an Airflow instance across multiple machines.

Note that this setup represents one possible configuration of Astronomer Core and uses optional tools such as systemd and PostgreSQL. After successfully starting Airflow for the first time, we recommend reviewing this configuration and adjusting it based on the functional requirements for your project.

If you haven't tested Airflow locally and would like to do so, refer to [Astronomer Core Quickstart](/docs/ac/next/01_quickstart).

## Prerequisites

First, ensure the OS-level packages listed below are installed on your machines:

- sudo
- python3
- python3-dev (python3-devel for RHEL/CentOS)
- python3-venv
- python3-psycopg2
- gcc
- postgresql
- systemd

> If you're Debian-based, run ` sudo apt-get install <package1> <package-2> ... <package-x>` to do so. If you're running RedHat Linux, run `$ yum install <package>`.

Reviewer: ```suggestion
If you're Debian-based, run `sudo apt-get install <package1> <package-2> ... <package-x>` to do so. If you're running RedHat Linux, run `yum install <package>`.
```

Separate question - Would we ever just give them this command so they don't have to type it out? So,

> If you're Debian-based, run `sudo apt-get install sudo python3  python3-dev python3-venv python3-psycopg2 gcc postgresql systemd` to install all packages with a single command. If you're running RedHat Linux, replace `sudo apt-get` with `yum install`.

Author: As with other long lists, I think they benefit more from seeing the prerequisites in bullets rather than in a single line. If the command was smaller, however, I'd definitely just write it out in a line.

Reviewer: Resolved over zoom

    ```sh
    sudo -u postgres createuser airflow -P
    ```

    This will prompt you for a password. Create one, and make a note of it for later.

2. Create a database named `airflow` and set the `airflow` user as the owner:

    ```sh
    sudo -u postgres createdb --owner airflow airflow
    ```

This guide assumes that your database server is local to where you run these commands and that you're on a Debian-like OS. If your setup is different, you will need to tweak these commands.

> **Note:** To make the database server accessible outside of your localhost, you may have to edit your [`/var/lib/postgresql/data/pg_hba.conf`](https://www.postgresql.org/docs/10/auth-pg-hba-conf.html) file and restart Postgres. Editing this file will vary for each individual database setup. Before editing this file, take a moment to assess the security  implications of editing this file.

### Alternative setup: Use an existing database

Instead of creating a new PostgreSQL database, you can use an existing database as long as both of the following are true:

- The database is compatible with Airflow as described in Prerequisites.
- A user named `airflow` has ownership access to the database.

> When you specify the `AIRFLOW__CORE__SQL_ALCHEMY_CONN` environment variable in step 2F, replace the connection URL with the appropriate URL for your database.

Reviewer: ```suggestion
When you specify the `AIRFLOW__CORE__SQL_ALCHEMY_CONN` environment variable in step 2F, replace the connection string with one that corresponds to your database.
```

So I _believe_ that the value defined by `AIRFLOW__CORE__SQL_ALCHEMY_CONN` is technically called a "connection string", per Airflow docs: https://airflow.apache.org/docs/apache-airflow/1.10.15/best-practices.html?highlight=sql_alchemy_conn#database-backend

## Step 2: Configure Each Machine in Your System

> Complete the steps below for each machine that will be running a core Apache Airflow component (Scheduler, Webserver, or Worker).

Reviewer: Should we add, "If you're running all components on a single VM, you only have to follow these steps once."

Author: I think that's implied by the instructions.

### A. Create a system user to run Airflow

Airflow can run as any user, but for this setup we assume the user name `astro`. Run the following command to add this user to your machine:

```sh
sudo useradd --create-home astro
```

### B. Create an Airflow project directory

You also need to configure an `AIRFLOW_HOME` directory (not to be confused with the user's home directory) where you'll store your DAGs. We recommend using the path `/usr/local/airflow` as your project directory and `/usr/local/airflow/dags` as your DAG directory, but any path can be chosen as long as the `astro` user has write access to it. To do this, run the following commands:

```sh
sudo install --owner=astro --group=astro -d /usr/local/airflow
echo 'export AIRFLOW_HOME=/usr/local/airflow' | sudo tee --append ~astro/.bashrc
cd /usr/local/airflow
sudo mkdir dags
```

> > **Note:** If you're running Airflow on multiple machines, each machine needs access to the same DAGs in order to successfully execute them. We recommend setting up automation pipelines for updating all of your DAG folders whenever a local folder is updated. For more information, read [Deploying DAGs].

Reviewer: Update this `[Deploying DAGs]` URL since that file is live?

    AIRFLOW__WEBSERVER__BASE_URL=http://localhost:8080
    AIRFLOW__CORE__EXECUTOR=LocalExecutor
    ```

- For Celery Executor:

    ```
    AIRFLOW__CORE__SQL_ALCHEMY_CONN=postgresql+psycopg2://airflow:<your-user-password>@localhost/airflow
    AIRFLOW__WEBSERVER__BASE_URL=http://localhost:8080
    AIRFLOW__CELERY__BROKER_URL=sqla+postgresql://airflow:<your-user-password>@localhost/airflow
    AIRFLOW__CELERY__RESULT_BACKEND=db+postgresql://airflow:<your-user-password>@localhost/airflow
    AIRFLOW__CORE__EXECUTOR=CeleryExecutor
    ```

The password you specify here should be the same one you specified when prompted by the `createuser` command in Step 1. If your password contains `%`, `/`, or `@` then you will need to url-escape; replace `%` with `%25`, `/` with `%2F`, and `@` with `%40`.

When you've finished configuring environment variables, run the following command to add your environment variables to your `astro` user's shell environment:

```sh
echo 'set -a; source /usr/local/airflow/env-vars; set +a' | sudo tee --append ~astro/.bashrc
```

#### Alternative setup: Database access

> * Your Airflow user password is stored in your `env-vars` file (owned by `root:root` and `0600` permissions) on your nodes. If you'd rather use an existing credential store, such as [HashiCorp Vault](https://www.hashicorp.com/products/vault), you can instead specify a command that will be run (once, at service start up) to obtain the connection string. For example:

Reviewer: ```suggestion
* Your Airflow user password is stored in your `env-vars` file (owned by `root:root` and `0600` permissions) on your nodes. If you'd rather use an existing credential store, such as [HashiCorp Vault](https://www.hashicorp.com/products/vault), you can instead specify a command to obtain the connection string. That command will run when the service starts up. For example:
```

    ```
    AIRFLOW__CORE__SQL_ALCHEMY_CONN_CMD=vault kv get -field=dsn secret/airflow-db
    ```

    For more information on this feature, read [Integrating Airflow and Hashicorp Vault](/guides/airflow-and-hashicorp-vault).

> * In this example setup, the `AIRFLOW__CELERY__BROKER_URL` environment variable uses the main database for communication. To take some load off of your main database, we recommend using a dedicated message broker such as [Redis](https://redis.io/) and specifying that here instead.

Reviewer: So, the format for setting the redis Env Var is `AIRFLOW__CELERY__BROKER_URL: redis://:@redis:6379/0` (at least per our [docker-compose files](https://github.com/astronomer/docker-airflow/blob/main/docker-compose-postgres-celery.yaml#L18)). A few questions as I'm thinking through this:

1. If we assume `redis` for docker-compose, should we just have users install + use that here since it's best practice? I.e. make it more prominent?
2. If not to above, should we at least give an example in this section of what that Env Var for redis should look like, and what else they have to do/install?
3. I'm slightly confused by the `#### Alternative setup: Database access` header - that makes me think that if I did "Alternative setup" above, that I'd follow these steps instead, but not otherwise. Should we rename to "Additional Database Considerations", or something like that?

## Step 3: Set Up the Scheduler

In Airflow, [the Scheduler](https://airflow.apache.org/docs/apache-airflow/stable/scheduler.html) is responsible for reading from the metadata database to check on the status of each task and decides the order in which tasks should be completed. To get your Scheduler running:  

1. Enable the Scheduler by running the following command:

    ```sh
    sudo systemctl enable astronomer-core@scheduler.service
    ```

2. Edit the override file for the machine by running the following command:

    ```sh
    sudo systemctl edit astronomer-core@scheduler.service
    ```

> 3. In the empty document that appears, add the following lines:

Reviewer: ```suggestion
3. In the empty file generated by the command above, add the following lines:
```

"Document that appears" sounds ominous lol - where does it appear, and what's it called?

Author: Yeah I always use "appears" for UI elements but it definitely sounds out of place when talking about command line-level stuff.

    ```
    [Service]
    ExecStartPre=/home/astro/airflow-venv/bin/airflow db upgrade
    ```

4. Start the service by running:   

    ```sh
    sudo systemctl start astronomer-core@scheduler.service
    ```

## Step 4: Set Up the Webserver

[The Webserver](https://airflow.apache.org/docs/apache-airflow/stable/security/webserver.html) is a core Airflow component that is responsible for rendering the Airflow UI. To configure it on its own machine, follow the steps below.

1. Enable the Webserver by running the following:

    ```sh
    sudo systemctl enable astronomer-core@webserver.service
    ```

2. Start the Webserver by running the following:

    ```sh
    sudo systemctl start astronomer-core@webserver.service
    ```

> You can now access the Airflow UI in a web browser via `http://localhost:8080`.

Reviewer: At this point, a user can only really see the "login" screen, right? Should we specify that they can't expect to actually log in ye/have access to the full application?

> **Note:** For added security and stability, we recommend running the Webserver behind a reverse proxy and load balancer such as [nginx](https://www.nginx.com/). For more information on this feature, read the [Apache Airflow documentation](https://airflow.apache.org/docs/stable/howto/run-behind-proxy.html).

## Step 5: Set Up the Worker Machines

For each machine on which you want to host a Worker:

1. Enable the Worker service by running the following command:

   ```sh
   sudo systemctl enable astronomer-core@worker.service
   ```

2. Start the service by running the following command:

    ```sh
    sudo systemctl start astronomer-core@worker.service
    ```

You now have the ability to run Airflow tasks within DAGs.

## Step 6: Create a User

To log into the Airflow UI, you need to first create an Airflow user:

1. Switch to your system `astro` user using the following command:

    ```sh
    sudo su astro /bin/bash
    ```

    All Airflow CLI commands must be run from your `astro` user.

2. Create a new Admin Airflow user with the following command:

    ```sh
    airflow users create -e EMAIL -f FIRSTNAME -l LASTNAME -p PASSWORD -r Admin -u USERNAME
    ```

>     When you access the Airflow UI, use this user information to log in for the first time.

Reviewer: ```suggestion
    To log in to the Airflow UI for the first time at http://localhost:8080, enter the username and password you set above.
```

Author: Eh, I'm actually going to cut this sentence altogether. This info is used immediately in the next step, and I think there's a very clear connection between creating this user and logging in.

## Step 7: Confirm the Installation

To confirm that you successfully installed Apache Airflow, open `http://localhost:8080` in your web browser. You should see the login screen for the Airflow UI.

> Log in with your `Admin` user. From there, you should see Airflow's primary 'DAGs' view:

Reviewer: Slightly redundant with last 2 sentences above? Maybe pull out what I just suggested ("To log in to the Airflow UI for the first time at http://localhost:8080, enter the username and password you set above.") and throw down here/combine?

Author: Opted to keep this and remove the other sentence.

---
title: "Install Apache Airflow on a Virtual Machine"
navTitle: "Install on a Virtual Machine"
description: "Install Apache Airflow on one or more virtual machines with the Astronomer Core Python wheel."
---

## Overview

This guide walks through the necessary steps to install Apache Airflow via the Astronomer Core Python wheel on one or more Virtual Machines (VMs). By the end of the setup, you'll be able to deploy and run Airflow across multiple machines.

If you haven't tested Airflow locally and would like to do so, refer to [Astronomer Core Quickstart](/docs/ac/next/01_quickstart).

## Prerequisites

> First, ensure the OS-level packages listed below are installed on your machines. If you're Debian-based, run ` sudo apt-get install <package>` to do so. If you're running RedHat Linux, run `$ yum install <package>`.

Reviewer: @jwitz I think we found that users can install these packages with a single command. Want to pull down "If you're Debian-based..." to the bottom of this package list, and say something like, 

> If you're Debian-based, run `sudo apt-get install <package-1> <package-2> <package-3>` to install the packages above with a single command. If you're running RedHat Linux, run `$ yum install <package-1> <package-2> <package-3>`.

- sudo
- python3
- python3-dev (python3-devel for RHEL/CentOS)
- python3-venv
- python3-psycopg2
- gcc
- postgresql
- systemd

You also need a database that is accessible to all the machines that will run your Airflow instance. While this guide walks through the process for configuring a PostgreSQL database, Airflow is compatible with all of the following databases:

- PostgreSQL: 9.6, 10, 11, 12, 13
- MySQL: 5.7, 8

> - SQLite: 3.15.0+

Reviewer: We list `postgresql` in the initial list of packages, but then we have the note below about database requirements. Maybe bring these sections together, or more clearly separate them somehow so people don't inadvertently install `postgresql` and then go try to install `SQLite`?

The rest of this doc mostly assumes postgres - maybe make that clear from the start?

Author: Brought the note up to this section, but I don't think this is the place to mention more about our PostgreSQL-biased setup. I'm going to include a note in the Overview that explains this is just one possible setup, and that you might want to configure things after getting started to better suit specific needs.

Lastly, you will need to run the following three Airflow components:

- Scheduler
- Webserver
- Worker(s)

You can run these components on one or multiple machines, though we recommend using multiple machines for a production environment.

> > **Note:** MySQL 5.7 is compatible with Airflow, but is not recommended for users running Airflow 2.0+, as it does not support the ability to run more than 1 Scheduler. If you'd like to leverage Airflow's new Highly-Available Scheduler, make sure you're running MySQL 8.0+.

Reviewer: ```suggestion
> **Note:** MySQL 5.7 is compatible with Airflow, but is not recommended for users running Airflow 2.0+, as it does not support the ability to run more than 1 Scheduler. If you'd like to leverage Airflow's new [Highly-Available Scheduler](https://www.astronomer.io/blog/airflow-2-scheduler), make sure you're running MySQL 8.0+.
```

When you specify the `AIRFLOW__CORE__SQL_ALCHEMY_CONN` environment variable in step 2F, replace the connection URL with the appropriate URL for your database.

## Step 2: Configure Each Machine in Your System

Complete the steps below for each machine that will be running a core Apache Airflow component (Scheduler, Webserver, or Worker).

### A. Create a system user to run Airflow

Airflow can run as any user, but for this setup we assume the user name `astro`. Run the following command to add this user to your machine:

```sh
sudo useradd --create-home astro
```

### B. Create an Airflow project directory

You also need to configure an `AIRFLOW_HOME` directory (not to be confused with the user's home directory) where you'll store your DAGs. We recommend using the path `/usr/local/airflow` as your project directory and `/usr/local/airflow/dags` as your DAG directory, but any path can be chosen as long as the `astro` user has write access to it. To do this, run the following commands:

```sh
sudo install --owner=astro --group=astro -d /usr/local/airflow
echo 'export AIRFLOW_HOME=/usr/local/airflow' | sudo tee --append ~astro/.bashrc

> cd /usr/local/airflow && sudo mkdir dags

Reviewer: Separate out `&& sudo mkdir dags` into its own command. I'd just remove that line and add,

> Then, run:
> ```
> sudo mkdir dags
>```

Author: I think it makes sense to keep these commands coupled in the same code block, but I agree with removing the `&&`.

```

venv is a tool to create lightweight, isolated Python environments without affecting systemwide configuration. For more information, read [Python's venv documentation](https://docs.python.org/3/library/venv.html).

### D. Install Astronomer Core

Install the AC Python wheel onto your machine by running:

```sh
sudo -u astro ~astro/airflow-venv/bin/pip install --extra-index-url=https://pip.astronomer.io/simple/ 'astronomer-core[postgres]==<airflow-version>' --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-<airflow-version>/constraints-3.8.txt"
```

To install the latest patch version of Apache Airflow 2.0.1, for example, this command would be:

```sh
sudo -u astro ~astro/airflow-venv/bin/pip install --extra-index-url=https://pip.astronomer.io/simple/ 'astronomer-core[postgres]==2.0.1.*' --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-2.0.1/constraints-3.8.txt"
```

This command includes the optional `[postgres]` dependency so that all libraries needed to use Postgres are also installed. If you are using a different database or require additional dependencies, specify those dependencies in a comma-delimited list:

```

> astronomer-core[mysql, redis, crypto, aws, celery]==2.0.1.*

Reviewer: If they choose to do this setup for the Celery Executor, do you know if they have to install `celery` here?

Author: Yep, tried to make this clearer in the sentence leading into the snippet.

To use systemd as a process supervisor:

1. Create a systemd unit file using the following command:

    ```sh
    sudo -e /etc/systemd/system/astronomer-core@.service
    ```

2. Using a text editor, create a file called `env-vars` and edit it to contain these environment variables and values:

    ```sh
    AIRFLOW_HOME=/usr/local/airflow/
    AIRFLOW__CORE__LOAD_EXAMPLES=False
    PATH=$PATH:/home/astro/airflow-venv/bin
    ```

    When you run Airflow for the first time, a file called `airflow.cfg` will be generated in your `AIRFLOW_HOME` directory. If you want to configure environment variables that apply to all of your machines, we recommend specifying them in that `airflow.cfg` file. For more information, read the Apache Airflow documentation on [Setting Configuration Options](https://airflow.apache.org/docs/apache-airflow/stable/howto/set-config.html).

3. Add the following to your systemd unit file:

    ```
    [Unit]
    Description=Airflow %I daemon
    After=network-online.target cloud-config.service
    Requires=network-online.target

    [Service]

>     EnvironmentFile=/etc/default/env-vars

Reviewer: I have a note from last week that we should:

- Specify `EnvironmentFile=/path/to/file/default/astronomer-core` (this in your airflow project directory)

You'll remember this better than I - is there a change needed here?

Author: I changed this back because I realized that `etc/default` is an actual filepath on Linux that people install things to. I guess the question is now, do we want to have this file in the Airflow project folder or elsewhere? Given that we have an `airflow-settings` file in the quickstart and local installation, I now think we should keep this in our project folder - updating accordingly.

    User=astro
    Group=astro
    Type=simple
    WorkingDirectory=/usr/local/airflow/
    ExecStart=/home/astro/airflow-venv/bin/airflow %i
    Restart=always
    RestartSec=5s

    [Install]
    WantedBy=multi-user.target
    ```

### F. Configure Airflow for Database Access

To connect your Airflow environment to the metadata DB you created in Step 1, add the following environment variables to your `env-vars` file depending on your chosen [Executor](https://www.astronomer.io/guides/airflow-executors-explained):

- For Local Executor:

    ```
    AIRFLOW__CORE__SQL_ALCHEMY_CONN=postgresql+psycopg2://airflow:<your-user-password>@localhost/airflow
    AIRFLOW__WEBSERVER__BASE_URL=http://localhost:8080
    AIRFLOW__CORE__EXECUTOR=LocalExecutor
    ```

- For Celery Executor:

    ```
    AIRFLOW__CORE__SQL_ALCHEMY_CONN=postgresql+psycopg2://airflow:<your-user-password>@localhost/airflow

>     AIRFLOW__WEBSERVER__BASE_URL=http://host:port

Reviewer: ```suggestion
    AIRFLOW__WEBSERVER__BASE_URL=http://localhost:8080
```
